tests/unit/models/keras/test_sampler.py

```diff
@@ -18,71 +18,221 @@
+        1e-1, tf.reduce_max(tf.abs(trajectory1(test_data)[:, 0] - trajectory1(test_data)[:, 1]))
+    )  # distinct for two samples within same draw
+    npt.assert_array_less(
+        1e-1, tf.reduce_max(tf.abs(trajectory2(test_data)[:, 0] - trajectory2(test_data)[:, 1]))
+    )  # distinct for two samples within same draw
 
 
 @random_seed
-def test_ensemble_trajectory_sampler_samples_are_distinct_for_new_instances() -> None:
+def test_ensemble_trajectory_sampler_samples_are_distinct_within_batch(
+    use_samples: bool,
+) -> None:
+    """
+    Samples for elements of the batch should be different. Note that when use_samples is not used,
+    for small ensembles we could randomnly choose the same network and then we would get the same
+    result.
+    """
     example_data = empty_dataset([1], [1])
     test_data = tf.linspace([-10.0], [10.0], 100)
     test_data = tf.expand_dims(test_data, -2)  # [N, 1, d]
+    test_data = tf.tile(test_data, [1, 2, 1])  # [N, 2, D]
 
-    model, _, _ = trieste_deep_ensemble_model(example_data, _ENSEMBLE_SIZE * 2)
+    model, _, _ = trieste_deep_ensemble_model(example_data, _ENSEMBLE_SIZE * 3)
 
-    sampler1 = EnsembleTrajectorySampler(model)
+    sampler1 = EnsembleTrajectorySampler(model, use_samples=use_samples)
     trajectory1 = sampler1.get_trajectory()
 
-    sampler2 = EnsembleTrajectorySampler(model)
+    sampler2 = EnsembleTrajectorySampler(model, use_samples=use_samples)
     trajectory2 = sampler2.get_trajectory()
 
-    assert tf.reduce_any(trajectory1(test_data) != trajectory2(test_data))
+    eval_1 = trajectory1(test_data)
+    eval_2 = trajectory2(test_data)
+
+    npt.assert_allclose(eval_1, eval_2)  # same for seperate draws
+    npt.assert_array_less(
+        1e-1, tf.reduce_max(tf.abs(eval_1[:, 0] - eval_1[:, 1]))
+    )  # distinct for two samples within same draw
+    npt.assert_array_less(
+        1e-1, tf.reduce_max(tf.abs(eval_2[:, 0] - eval_2[:, 1]))
+    )  # distinct for two samples within same draw
 
 
 @random_seed
-def test_ensemble_trajectory_sampler_resample_provides_new_samples_without_retracing() -> None:
+def test_ensemble_trajectory_sampler_resample_with_new_sampler_does_not_change_old_sampler(
+    use_samples: bool,
+) -> None:
+    """
+    Generating a new sampler and resampling trajectories in it will not affect a previous
+    sampler instance. Before resampling evaluations from trajectories of both samplers
+    are the same.
+    """
+    example_data = empty_dataset([1], [1])
+    test_data = tf.linspace([-10.0], [10.0], 100)
+    test_data = tf.expand_dims(test_data, -2)  # [N, 1, d]
+    test_data = tf.tile(test_data, [1, 2, 1])  # [N, 2, D]
+
+    model, _, _ = trieste_deep_ensemble_model(example_data, _ENSEMBLE_SIZE * 3)
+
+    sampler1 = EnsembleTrajectorySampler(model, use_samples=use_samples)
+    trajectory1 = sampler1.get_trajectory()
+    evals_11 = trajectory1(test_data)
+
+    sampler2 = EnsembleTrajectorySampler(model, use_samples=use_samples)
```

Comment:
if I try to do this without creating another sampler old trajectory is changed as well, perhaps we can live with this but its not ideal (@uri-granta said it effectively means that sampler is changed when we change trajectory, not a desired behaviour...)